samples/bluetooth/broadcast_assistant/src/main.c

```diff
@@ -0,0 +1,415 @@
+
+	err = k_sem_take(&sem_sink_discovered, K_FOREVER);
+	if (err != 0) {
+		printk("Failed to take sem_sink_discovered (err %d)\n", err);
+	}
+}
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	(void)bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (err != 0) {
+		printk("Failed to connect to %s (%u)\n", addr, err);
+
+		bt_conn_unref(broadcast_sink_conn);
+		broadcast_sink_conn = NULL;
+
+		scan_for_broadcast_sink();
+		return;
+	}
+
+	if (conn != broadcast_sink_conn) {
+		return;
+	}
+
+	printk("Connected: %s\n", addr);
+	k_sem_give(&sem_sink_connected);
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	if (conn != broadcast_sink_conn) {
+		return;
+	}
+
+	(void)bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	printk("Disconnected: %s (reason 0x%02x)\n", addr, reason);
+
+	bt_conn_unref(broadcast_sink_conn);
+	broadcast_sink_conn = NULL;
+
+	k_sem_give(&sem_sink_disconnected);
+}
+
+static void security_changed_cb(struct bt_conn *conn, bt_security_t level,
+				enum bt_security_err err)
+{
+	if (err == 0) {
+		printk("Security level changed: %u\n", level);
+		k_sem_give(&sem_security_updated);
+	} else {
+		printk("Failed to set security level: %u\n", err);
+	}
+}
+
+static void reset(void)
+{
+	broadcast_sink_conn = NULL;
+	selected_broadcast_id = INVALID_BROADCAST_ID;
+	selected_sid = 0;
+	selected_pa_interval = 0;
+	(void)memset(&selected_addr, 0, sizeof(selected_addr));
+
+	k_sem_reset(&sem_source_discovered);
+	k_sem_reset(&sem_sink_discovered);
+	k_sem_reset(&sem_sink_connected);
+	k_sem_reset(&sem_sink_disconnected);
+	k_sem_reset(&sem_security_updated);
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected = connected,
+	.disconnected = disconnected,
+	.security_changed = security_changed_cb
+};
+
+int main(void)
+{
+	int err;
+	struct bt_bap_scan_delegator_subgroup subgroup = { 0 };
+	struct bt_bap_broadcast_assistant_add_src_param param = { 0 };
+
+	err = bt_enable(NULL);
+	if (err) {
+		printk("Bluetooth init failed (err %d)\n", err);
+		return 0;
+	}
+
+	printk("Bluetooth initialized\n");
+
+	bt_le_scan_cb_register(&scan_callbacks);
+
+start:
+	scan_for_broadcast_sink();
+
+	err = k_sem_take(&sem_sink_connected, K_FOREVER);
+	if (err != 0) {
+		printk("Failed to take sem_sink_connected (err %d)\n", err);
+	}
+
+	err = bt_bap_broadcast_assistant_discover(broadcast_sink_conn);
+	if (err != 0) {
+		printk("Failed to discover BASS on the sink (err %d)\n", err);
+	}
+
+	err = k_sem_take(&sem_security_updated, K_SECONDS(10));
+	if (err != 0) {
+		printk("Failed to take sem_security_updated (err %d), resetting\n", err);
+		bt_conn_disconnect(broadcast_sink_conn, BT_HCI_ERR_AUTH_FAIL);
+
+		if (k_sem_take(&sem_sink_disconnected, K_SECONDS(10)) != 0) {
+			/* This should not happen */
+			return -ETIMEDOUT;
+		}
+
+		reset();
+		goto start;
+	}
```

Comment:
I would add a TODO here to discover and parse the PACS on the sink, so we know what it supports